client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierConsumerAgent.java

```diff
@@ -0,0 +1,552 @@
+    this.bufferReaders = new HashMap<>();
+    this.receivedBuffers = new HashMap<>();
+    this.subPartitionsNeedNotifyAvailable = new HashSet<>();
+    for (TierShuffleDescriptor shuffleDescriptor : shuffleDescriptors) {
+      if (shuffleDescriptor instanceof TierShuffleDescriptorImpl) {
+        initShuffleClient((TierShuffleDescriptorImpl) shuffleDescriptor);
+        break;
+      }
+    }
+    checkNotNull(this.shuffleClient);
+    initBufferReaders();
+  }
+
+  @Override
+  public void setup(TieredStorageMemoryManager memoryManager) {
+    this.memoryManager = memoryManager;
+    for (Map<TieredStorageSubpartitionId, CelebornChannelBufferReader> subPartitionReaders :
+        bufferReaders.values()) {
+      subPartitionReaders.forEach((partitionId, reader) -> reader.setup(memoryManager));
+    }
+  }
+
+  @Override
+  public void start() {
+    // notify input gate that some sub partitions are available
+    Set<Tuple2<TieredStoragePartitionId, TieredStorageSubpartitionId>> needNotifyAvailable;
+    synchronized (lock) {
+      needNotifyAvailable = new HashSet<>(subPartitionsNeedNotifyAvailable);
+      subPartitionsNeedNotifyAvailable.clear();
+      started = true;
+    }
+    try {
+      needNotifyAvailable.forEach(
+          partitionIdTuple -> notifyAvailable(partitionIdTuple.f0, partitionIdTuple.f1));
+    } catch (Throwable t) {
+      LOG.error("Error occurred when notifying sub partitions available", t);
+      recycleAllResources();
+      ExceptionUtils.rethrow(t);
+    }
+    needNotifyAvailable.clear();
+
+    // Require segment 0 when starting the client
+    for (TieredStorageConsumerSpec spec : consumerSpecs) {
+      for (int subpartitionId : spec.getSubpartitionIds().values()) {
+        CelebornChannelBufferReader bufferReader =
+            getBufferReader(spec.getPartitionId(), new TieredStorageSubpartitionId(subpartitionId));
+        if (bufferReader == null) {
+          continue;
+        }
+        // TODO: if fail to open reader, may the downstream task start before than upstream task,
+        // should retry open reader, rather than throw exception
+        boolean openReaderSuccess = openReader(bufferReader);
+        if (!openReaderSuccess) {
+          LOG.error("Failed to open reader.");
+          recycleAllResources();
+          ExceptionUtils.rethrow(new IOException("Failed to open reader."));
+        }
+        bufferReader.notifyRequiredSegmentIfNeeded(0, subpartitionId);
+      }
+    }
+  }
+
+  @Override
+  public int peekNextBufferSubpartitionId(
+      TieredStoragePartitionId tieredStoragePartitionId,
+      ResultSubpartitionIndexSet resultSubpartitionIndexSet) {
+    synchronized (lock) {
+      // check health
+      healthCheck();
+
+      // return the subPartitionId if already receive buffer from corresponding subpartition
+      Map<TieredStorageSubpartitionId, Queue<Buffer>> subPartitionReceivedBuffers =
+          receivedBuffers.get(tieredStoragePartitionId);
```

Comment:
Could `receivedBuffers` use `ConcurrentHashMap`? Meanwhile, the lock blocker only locks health check. WDYT?

Reply:
I think the lock cost is acceptable in the current implementation, and it is convenient for developers to understand.